src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,358 @@
+        {
+            _tryTakeTimeout = tryTakeTimeout;
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _removeExpiredValues = removeExpiredValues;
+            _eventQueueTask = new Task(OnStart, options);
+            _eventQueueTask.Start();
+            if (_removeExpiredValues)
+                _ = RemoveExpiredValuesPeriodically(TimeSpan.FromSeconds(cleanUpIntervalInSeconds));
+        }
+
+        private void OnStart()
+        {
+            while (!_disposed)
+            {
+                try
+                {
+                    if (_eventQueue.TryTake(out var action, _tryTakeTimeout))
+                        action.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10900, ex));
+                }
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal void WaitForProcessing()
+        {
+            while (!_disposed)
+            {
+                if (_eventQueue.Count == 0)
+                    return;
+            }
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            try
+            {
+                var node = _doubleLinkedList.First;
+                while (node != null)
+                {
+                    var nextNode = node.Next;
+                    if (node.Value.ExpirationTime < DateTime.UtcNow)
+                    {
+                        _doubleLinkedList.Remove(node);
+                        _map.TryRemove(node.Value.Key, out _);
+                        numItemsRemoved++;
+                    }
+
+                    node = nextNode;
+                }
+            }
+            catch (ObjectDisposedException ex)
+            {
+                LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10902, nameof(RemoveExpiredValues), ex));
+            }
+
+            return numItemsRemoved;
+        }
+
+        internal void RemoveLRUs()
+        {
+            // use the _capacity for the newCacheSize calculation in the case where the cache is experiencing overflow
+            int currentCount = _map.Count <= _capacity ? _capacity : _map.Count;
+            var newCacheSize = currentCount - (int)(currentCount * _compactionPercentage);
+            while (_map.Count > newCacheSize && _doubleLinkedList.Count > 0)
+            {
+                var lru = _doubleLinkedList.Last;
```

Comment:
Is it possible at this time that _doubleLinkedList.Last is null?
I guess it hinges if RemoveExpiredValues runs parallel to this task.
A race condition could have the last item removed between the check for _doubleLinkedList.Count > 0 and var lru = _doubleLinkedList.Last.

Reply:
RemoveLRUs and RemoveExpiredValues are both processed by the _eventQueue, so they should only ever run sequentially one after the other.